### PR TITLE
VZ-9205-Decouple node drivers and cloud creds

### PR DESCRIPTION
**Pull request by theabmitra**

### CODE REVIEW

File: pkg/data/management/machinedriver_data.go [hunk 1]
```diff
@@ -39,3 +49,100 @@ var DriverData = map[string]map[string][]string{
...
+	addCloudCredential bool
+	url                string
+	uiURL              string
+	checksum           string
+	name               string
+	whitelist          []string
+	annotations        map[string]string
+}
+
+func addMachineDrivers(management *config.ManagementContext) error {
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go]: no oci driver?

File: package/Dockerfile [hunk 2]
```diff
@@ -114,6 +114,7 @@ RUN mkdir -p /var/lib/rancher/k3s/agent/images/ && \
     mkdir -p /var/lib/rancher-data/driver-metadata && \
     mkdir -p /var/log/auditlog
 
+RUN curl -sLf https://github.com/rancher/machine/releases/download/${CATTLE_MACHINE_VERSION}/rancher-machine-${ARCH}.tar.gz | tar xvzf - -C /usr/bin
```
Comment (anders-swanson) [inline package/Dockerfile]: do we still need to download this?

File: pkg/data/management/machinedriver_data.go [hunk 3]
```diff
@@ -18,24 +33,286 @@ const (
...
+
+	if err := dc.addCloudCredential(GoogleDriver); err != nil {
+		return err
+	}
+	return dc.addCloudCredential(OCIDriver)
+}
+
+func (d *DynamicSchemaClients) addCloudCredential(name string) error {
+	// annotations can have keys cred and password, values []string to be considered as a part of cloud credential
+	credLock.Lock()
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go]: what is the lock for?

File: pkg/data/management/machinedriver_data.go [hunk 4]
```diff
@@ -18,24 +33,286 @@ const (
...
+	defer credLock.Unlock()
+
+	err := errs.New("cloud credential creation failed")
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go]: Why is err being assigned here?

File: pkg/data/management/machinedriver_data.go [hunk 5]
```diff
@@ -18,24 +33,286 @@ const (
...
+	var existingSchema *v3.DynamicSchema
+	annotations := map[string]string{}
+	if existingSchema != nil {
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go]: existingSchema is always nil here? You define it on L111

File: pkg/data/management/machinedriver_data.go [hunk 6]
```diff
@@ -18,24 +33,286 @@ const (
...
+	}
+
+	flags, err := utils.GetCreateFlagsForDriver(name)
+	if err != nil {
+		return err
+	}
+	credFields := map[string]v32.Field{}
+	resourceFields := map[string]v32.Field{}
+
+	pubCredFields, privateCredFields, passwordFields, defaults, optionals := utils.GetCredFields(annotations)
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go:114]: consider returning a struct when you have more than 2 return values

File: pkg/data/management/machinedriver_data.go [hunk 7]
```diff
@@ -18,24 +33,286 @@ const (
...
+	}
+	dynamicSchema := &v3.DynamicSchema{
+		Spec: v32.DynamicSchemaSpec{
+			ResourceFields: resourceFields,
+		},
+	}
+
+	// Creating dynamic schema cloud config objects
+	dynamicSchema.Name = name + "config"
+	//dynamicSchema.OwnerReferences = []metav1.OwnerReference{
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go]: remove commented code

File: pkg/data/management/machinedriver_data.go [hunk 8]
```diff
@@ -18,24 +33,286 @@ const (
...
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			credentialSchema := &v3.DynamicSchema{
+				Spec: v32.DynamicSchemaSpec{
+					ResourceFields: credFields,
+				},
+			}
+			credentialSchema.Name = name
+			//credentialSchema.OwnerReferences = []metav1.OwnerReference{
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go]: remove commented code

File: pkg/data/management/machinedriver_data.go [hunk 9]
```diff
@@ -18,24 +33,286 @@ const (
...
+		_, err := d.schemaClient.Update(toUpdate)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *DynamicSchemaClients) createOrUpdateNodeForEmbeddedTypeCredential(embeddedType, fieldName string, embedded bool) error {
+	credSchemaLock.Lock()
```
Comment (anders-swanson) [inline pkg/data/management/machinedriver_data.go]: is there some race condition here?

File: pkg/data/management/utils/configData.go [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+			&cli.StringFlag{
+				Name: "amazonec2-session-token",
+			},
+			&cli.StringFlag{
+				Name: "amazonec2-ami",
+			},
+			&cli.StringFlag{
+				Name: "amazonec2-region",
+			},
+			&cli.StringFlag{
```
Comment (anders-swanson) [inline pkg/data/management/utils/configData.go]: why do we have all these extra amazon fields? These are all a part of the amazon cloud credential?

File: pkg/data/management/utils/configData.go [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+				Name: "amazonec2-http-endpoint",
+			},
+			&cli.StringFlag{
+				Name: "amazonec2-http-tokens",
+			},
+		}
+
+		/*
+			(*mcnflag.StringFlag)(0xc0024a1600)(azure-environment),
```
Comment (anders-swanson) [inline pkg/data/management/utils/configData.go]: remove commented code

File: pkg/data/management/utils/configData.go [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+			 (*mcnflag.BoolFlag)(0xc002340de0)(azure-no-public-ip),
+			 (*mcnflag.BoolFlag)(0xc0023411a0)(azure-static-public-ip),
+			 (*mcnflag.StringFlag)(0xc0024a1c40)(azure-dns),
+			 (*mcnflag.StringSliceFlag)(0xc0084ec280)(azure-open-port),
+			 (*mcnflag.StringFlag)(0xc0024a1d40)(azure-client-id),
+			 (*mcnflag.StringFlag)(0xc0024a1d80)(azure-client-secret)
+		*/
+	case "azure":
+		flags = []cli.Flag{
+			&cli.StringFlag{
```
Comment (anders-swanson) [inline pkg/data/management/utils/configData.go:32]: these look like full driver flags, not cloud credential flags

File: pkg/data/management/utils/utils.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+		field.Default.StringValue = strconv.Itoa(v.Value)
+	case *cli.BoolFlag:
+		field.Type = "boolean"
+		field.Description = v.Usage
+	case *cli.StringSliceFlag:
+		field.Type = "array[string]"
+		field.Description = v.Usage
+		field.Nullable = true
+		field.Default.StringSliceValue = v.Value
+	case *BoolPointerFlag:
```
Comment (anders-swanson) [inline pkg/data/management/utils/utils.go:60]: Is this ever the case? I don't see utils.BoolPointerFlag being created anywhere

Comment (anders-swanson) [reply pkg/data/management/utils/utils.go:60]: Or is this rancher code added back?

File: pkg/data/management/machinedriver_data.go (hunk 3 above)
Comment (theabmitra) [reply pkg/data/management/machinedriver_data.go]: multiple replicas of rancher will do the same thing .

File: pkg/data/management/utils/configData.go (hunk 10 above)
Comment (theabmitra) [reply pkg/data/management/utils/configData.go]: i kept this thinking if upgrade will be affected. without this our use case is working

File: pkg/data/management/machinedriver_data.go (hunk 5 above)
Comment (theabmitra) [reply pkg/data/management/machinedriver_data.go]: removed

File: pkg/data/management/machinedriver_data.go (hunk 3 above)
Comment (anders-swanson) [reply pkg/data/management/machinedriver_data.go]: it's in memory lock though, will not work across replicas?

File: pkg/data/management/utils/configData.go (hunk 10 above)
Comment (anders-swanson) [reply pkg/data/management/utils/configData.go]: what's the significance for upgrade?

**Review by anders-swanson — APPROVED** [review]

**Pull request merged**